---
title: "20440-Project"
author: "Christine Davis"
date: "3/29/2021"
output: html_document
---

```{r setup, include=FALSE}
rm(list = ls()) #remove stored variables
knitr::opts_chunk$set(echo = TRUE) #set knitr options
```

#Load packages

```{r load, include=FALSE}
library(readr)
library(pheatmap)
library(dendsort)
library(viridis)
set.seed(1234)  # random seed for reproducibility
```

#Download and clean data

```{r GEO data from downloaded csv}
# Import the data and look at the first six rows- gives data in log2 cpm
data <- read.csv(file = '../20440-Project/Data/ABC_WT_DKO.csv')
head(data)

# Pull out gene names, counts (log2 cpm)
genes <- data[,1]
counts <- data[,7:10]
rownames(counts) <- genes
counts <- t(counts)

#Z-score gene counts, create list of sample types
X <- scale(counts, center = TRUE, scale = TRUE)
ID <- c("WT","WT","DKO","DKO")
```

# Create heatmap for 2 page report

```{r Heatmap}

#Only use top-50 most significant differentially expressed genes
Xtmp <- X[,1:50]

#Cluster genes using hierarchical clustering with average Euclidean distance
mat_cluster_cols <- hclust(dist(t(Xtmp)), method="average")

#Define sorting function to sort dendrogram/clustering distances by magnitude
sort_hclust <- function(...) as.hclust(dendsort(as.dendrogram(...)))

#Sort columns and rows according to sorting function
mat_cluster_cols <- sort_hclust(mat_cluster_cols)
mat_cluster_rows <- sort_hclust(hclust(dist(X)))

#Define quantile_breaks function to take given matrix, split into 10 equal
#pieces, and return break locations. This will be used to color the heatmap
#with each color representing 10% of the data, to distinguish deciles
quantile_breaks <- function(xs, n = 10) {
  breaks <- quantile(xs, probs = seq(0, 1, length.out = n), na.rm = TRUE)
  breaks[!duplicated(breaks)]
}

#Use break function to create breaks in gene expression data 
mat_breaks <- quantile_breaks(Xtmp, n = 10)

#Label heatmap rows with sample type, columns with gene names
annotation_row = data.frame(strain = ID)
rownames(annotation_row) <- rownames(Xtmp)

#Provide colors for sample types
my_colors <- list(
  strain = c("WT" = '#8C02CC', "DKO" = '#5192C3'))

#generate heatmap of log2 cpm count for top-50 differentially expressed genes,
#using sample color assignment, clustering columns and rows in previously sorted 
#order from sort_hclust, breaks from previously defined mat_breaks,
#and providing specific font and cell sizes

plt <- pheatmap(Xtmp, color = viridis(9), 
         annotation_colors = my_colors, 
         annotation_row = annotation_row, 
         cellwidth = 8, 
         cellheight = 18, 
         fontsize = 12, 
         border_color = "grey", 
         treeheight_row = 3, 
         treeheight_col = 12,
         breaks = mat_breaks,
         cluster_cols = mat_cluster_cols,
         cluster_rows = mat_cluster_rows,
         cex = 0.7
         )

#Print heatmap to pdf
pdf(paste("../20440-Project/Figures/Pheatmap",".pdf", sep = ""), 
    width = 10, height = 8) 
print(plt)
dev.off()
```

