---
title: "20440-Project"
author: "Christine Davis"
date: "3/29/2021"
output: html_document
---

```{r setup, include=FALSE}
rm(list = ls()) #remove stored variables
knitr::opts_chunk$set(echo = TRUE) #set knitr options
```

#Load packages

```{r load, include=FALSE}
library(readr)
library(pheatmap)
library(dendsort)
library(viridis)
library(ggrepel)
library(ggplot2)
library(dplyr)
library(limma)
library(BiocManager)
#BiocManager::install("biomaRt")
library(dnapath)
set.seed(1234)  # random seed for reproducibility
```

#Download and clean data

```{r GEO data from downloaded csv}
# Import the data and look at the first six rows- gives data in log2 cpm
data <- read.csv(file = '../Data/ABC_WT_DKO.csv')
head(data)

# Pull out gene names, counts (log2 cpm)
genes <- data[,1]
counts <- data[,7:10]
rownames(counts) <- genes
counts <- t(counts)

#Z-score gene counts, create list of sample types
X <- scale(counts, center = TRUE, scale = TRUE)
ID <- c("WT","WT","DKO","DKO")

#Repeat for human data
hdata <- read.csv(file = '../Data/Human_SLE_Ctrl.csv')
head(hdata)

# Pull out gene names, counts (log2 cpm)
hgenes <- hdata[,1:2]
hcounts <- hdata[,4:ncol(hdata)]
rownames(hcounts) <- hgenes[,1]
hcounts <- t(hcounts)

#Z-score gene counts, create list of sample types
hX <- scale(hcounts, center = TRUE, scale = TRUE)

#Name samples by disease state
df_diagnosis <- data.frame(sample = rownames(hX))
df_diagnosis$diagnosis <- rep(NA, length = nrow(df_diagnosis))
df_diagnosis$diagnosis[which(grepl("HC", df_diagnosis$sample))] <- "Control"
df_diagnosis$diagnosis[which(grepl("SLE", df_diagnosis$sample))] <- "SLE"
hID <- df_diagnosis$diagnosis


```

#PCA for visualization of samples

```{r PCA}

#Create pca model from Z-scored gene counts
pca <- prcomp(hX)

## Add sample info to PCs and plot
labeled_PCs <- as.data.frame(cbind(as.character(hID), pca$x))
plt <- ggplot(labeled_PCs, aes(x = PC1, y=PC2, color=hID)) + 
  geom_point(shape = 19, size = 4) +
  ggtitle("PCA of WT vs DKO ABC Transcriptome") +
  theme_classic() +
  theme(axis.text.x = element_blank(), axis.text.y = element_blank(),
        axis.ticks = element_blank(), 
        plot.title = element_text(hjust=0.5, size = 18))

#Print PCA to pdf
pdf(paste("../Figures/PCA",".pdf", sep = ""), 
    width = 8, height = 8) 
print(plt)
dev.off()

```

# Differential Expression Analysis ("DE")

```{r Differential Expression}

#Create identification matrix designating each sample as DKO or WT
matrix <- model.matrix(~0+ID)
colnames(matrix) <- c("DKO","WT")

#Fit a linear model of gene counts by the design matrix, 
#so each group is compared. Transpose cpm counts to align with design matrix.
fit <- lmFit(t(counts), matrix)

#makeContrasts takes expression specifying contrasts, returns contrast matrix
contrasts <- makeContrasts(DKO - WT, levels=matrix)
#Compute estimated coefficients from linear model fit, contrasts
fit2 <- contrasts.fit(fit, contrasts)

#Takes microarray linear fit model, returns t-statistics, F-statistics, and 
#log-odds of differential expression using empirical Bayes' moderation
fit2 <- eBayes(fit2)

#Check how many genes are significantly over-expressed or under-expressed
#between conditions, and which are equally expressed (~10% of genes are 
#significantly diff. expressed)
table(decideTests(fit2)) #returns # of over (+1) or under (-1) expressed genes
#Get table of top-ranked genes, make gene names explicit variable
full_results <- topTable(fit2, number=Inf, adjust.method="BH")
full_results <- tibble::rownames_to_column(full_results,"ID")

p_cutoff <- 0.01
fc_cutoff <- 1
topN <- 20 #top N genes for annotation

#Reformat full results to have additional variables of interest, filter for 
#set parameters, plot volcano plot of DE and significance
full_results <- full_results %>% 
  mutate(Significant = adj.P.Val < p_cutoff, abs(logFC) > fc_cutoff ) %>% 
  mutate(Rank = 1:n(), Label = ifelse(Rank < topN, colnames(counts),""))
plt <- ggplot(full_results, aes(x = logFC, y = B, col=Significant,label=Label)) + geom_point() + 
  geom_text_repel(col="black", max.overlaps = 30) +
  ggtitle("Differential Expression Analysis- WT vs DKO ABCs") +
  xlab("Log fold-change") + ylab("Log odds of differential expression") +
  theme_light() +
  theme(plot.title = element_text(hjust=0.5, size = 18))  

#Print volano to pdf
pdf(paste("../Figures/Volcano",".pdf", sep = ""), 
    width = 10, height = 8) 
print(plt)
dev.off()

```

# Create heatmap for 2 page report

```{r Heatmap}

#Temporarily only use top-100 most significant differentially expressed genes
Xtmp <- X[,1:100]

#Cluster genes using hierarchical clustering with average Euclidean distance
order_cluster_cols <- hclust(dist(t(Xtmp)), method="average")

#Define sorting function to sort dendrogram/clustering distances by magnitude
sort_hclust <- function(...) as.hclust(dendsort(as.dendrogram(...)))

#Sort columns and rows according to sorting function
mat_cluster_cols <- sort_hclust(order_cluster_cols)
mat_cluster_rows <- sort_hclust(hclust(dist(X)))

#Define quantile_breaks function to take given matrix, split into 10 equal
#pieces, and return break locations. This will be used to color the heatmap
#with each color representing 10% of the data, to distinguish deciles
quantile_breaks <- function(xs, n = 10) {
  breaks <- quantile(xs, probs = seq(0, 1, length.out = n), na.rm = TRUE)
  breaks[!duplicated(breaks)]
}

#Use break function to create breaks in gene expression data 
mat_breaks <- quantile_breaks(Xtmp, n = 10)

#Label heatmap rows with sample type, columns with gene names
annotation_row = data.frame(strain = ID)
rownames(annotation_row) <- rownames(Xtmp)

#Provide colors for sample types
my_colors <- list(
  strain = c("WT" = '#8C02CC', "DKO" = '#5192C3'))

#generate heatmap of log2 cpm count for top-100 differentially expressed genes,
#using sample color assignment, clustering columns and rows in previously sorted 
#order from sort_hclust, breaks from previously defined mat_breaks,
#and providing specific font,cell, and dedrogram (tree) sizes

plt <- pheatmap(Xtmp, color = viridis(9), 
         annotation_colors = my_colors, 
         annotation_row = annotation_row, 
         cellwidth = 8, 
         cellheight = 18, 
         fontsize = 12, 
         border_color = "grey", 
         treeheight_row = 3, 
         treeheight_col = 12,
         breaks = mat_breaks,
         cluster_cols = mat_cluster_cols,
         cluster_rows = mat_cluster_rows,
         cex = 0.7
         )

#Print heatmap to pdf
pdf(paste("../Figures/Pheatmap",".pdf", sep = ""), 
    width = 20, height = 8) 
print(plt)
dev.off()
```

#Differential connectivity

```{r Package loading}

library(org.Hs.eg.db)
BiocManager::install("org.Mm.eg.db")
library(org.Mm.eg.db)
library(githubinstall)
githubinstall("dnapath")
library(dnapath)
BiocManager::install("reactome.db")
BiocManager::install("minet")
BiocManager::install("GENIE3")
library(reactome.db)
library(SeqNet)
```

``` {r DC analysis}
#Convert gene names to Entrez IDs
mouse_gene_names <- mapIds(org.Mm.eg.db, colnames(counts), 'ENTREZID', 'SYMBOL')
human_gene_names <- colnames(hcounts)

#Add entrez IDs to dataframe
mouse_Entrez_genes <- data.frame(counts)
colnames(mouse_Entrez_genes) <- mouse_gene_names
human_Entrez_genes <- data.frame(hcounts)
colnames(human_Entrez_genes) <- human_gene_names

#Get reactome pathways represented within genes measured for both mouse & human
react_paths_human <- get_reactome_pathways(species = "Homo sapiens")
react_paths_mouse <- get_reactome_pathways(species = "Mus musculus")

# Run dnapath using gene expression and group information for human data
results_human <- dnapath(human_Entrez_genes, 
                   pathway_list = react_paths_human, 
                   groups = df_diagnosis$diagnosis,
                   network_inference = run_pcor)

results_mouse <- dnapath(mouse_Entrez_genes, 
                   pathway_list = react_paths_mouse, 
                   groups = ID,
                   network_inference = run_corr)

results_human <- filter_pathways(results_human, alpha_pathway = 0.05)
results_mouse <- filter_pathways(results_mouse,alpha_pathway = 0.333) #limited by replicate number
# Setting seed allows for reproducible work
set.seed(123)
results_mouse <- rename_genes(results_mouse, to = "symbol", species = "Mus musculus",
                        dir_save = tempdir())
results_human <- rename_genes(results_human, to = "symbol", species = "Homo Sapiens",
                        dir_save = tempdir())

# Extract the two estimated association networks for the first pathway
nw <- get_networks(results_human[[1]])
# Plot the networks using the SeqNet::plot_network function.

# Plot the two networks (in separate plots)
g <- SeqNet::plot_network(nw[[1]])
SeqNet::plot_network(nw[[1]], compare_graph = g)
# Plot of the differential network for pathway 1.
# Use compare_graph to maintain same layout
plot(results_human[[1]], compare_graph = g)

# Investigate inflammasome and IFNG genes further
plot_pair(results_human[[1]], "42386", "13837") #Inflammasome vs IFNG

```

