---
title: "20440-Project"
author: "Christine Davis"
date: "3/29/2021"
output: html_document
---

```{r setup, include=FALSE}
rm(list = ls()) #remove stored variables
knitr::opts_chunk$set(echo = TRUE) #set knitr options
```

#Load packages

```{r load, include=FALSE}
library(readr)
library(pheatmap)
library(dendsort)
library(viridis)
library(ggrepel)
library(ggplot2)
library(dplyr)
library(limma)
library(BiocManager)
#BiocManager::install("biomaRt")
library(dnapath)
library(factoextra)
set.seed(1234)  # random seed for reproducibility
```

#Download and clean data !!!! SHOULD convert fpkm to tpm for final project

```{r GEO data from downloaded csv}
# Import the data and look at the first six rows- gives data in log2 cpm
data <- read.csv(file = '../Data/ABC_WT_DKO.csv')
head(data)

# Pull out gene names, counts (log2 cpm)
genes <- data[,1]
counts <- data[,7:10]
rownames(counts) <- genes
counts <- t(counts)
rev_log2 <- function(x) {( 2^x ) }  #(--> change back to regular cpm)
counts1 <- sapply(counts[1,],rev_log2)
counts2 <- sapply(counts[2,],rev_log2)
counts3 <- sapply(counts[3,],rev_log2)
counts4 <- sapply(counts[4,],rev_log2)
counts <- data.frame(rbind(counts1, counts2, counts3, counts4))
rownames(counts) <- colnames(data[,7:10])

#Z-score gene counts, create list of sample types
X <- scale(counts, center = TRUE, scale = TRUE)
ID <- c("WT","WT","DKO","DKO")

#Repeat for human data
hdata <- read.csv(file = '../Data/Human_SLE_Ctrl.csv')
head(hdata)

# Pull out gene names, counts (log2 cpm)
hgenes <- hdata[,1:2]
hcounts <- hdata[,4:ncol(hdata)]
rownames(hcounts) <- hgenes[,1]
hcounts <- hcounts[-which(is.na(hgenes[,2])),]
hgenes <- hgenes[-which(is.na(hgenes[,2])),]
hcounts <- hcounts[-which(grepl("2-Mar",hgenes[,2])),]
hgenes <- hgenes[-which(grepl("2-Mar",hgenes[,2])),]
rownames(hcounts) <- hgenes[,2]
hcounts <- t(hcounts)

#Z-score gene counts, create list of sample types
hX <- scale(hcounts, center = TRUE, scale = TRUE)

#Name samples by disease state
df_diagnosis <- data.frame(sample = rownames(hX))
df_diagnosis$diagnosis <- rep(NA, length = nrow(df_diagnosis))
df_diagnosis$diagnosis[which(grepl("HC", df_diagnosis$sample))] <- "Control"
df_diagnosis$diagnosis[which(grepl("SLE", df_diagnosis$sample))] <- "SLE"
hID <- df_diagnosis$diagnosis


```

#PCA for visualization of samples

```{r PCA}
#To run for mouse vs human, change X, ID (mouse) --> hX, hID (human)

#Create pca model from Z-scored gene counts
pca <- prcomp(hX)

## Add sample info to PCs and plot
labeled_PCs <- as.data.frame(cbind(as.factor(ID), pca$x))
plt <- ggplot(labeled_PCs, aes(x = PC1, y=PC2, color=ID)) + 
  geom_point(shape = 19, size = 10, alpha = 0.8) +
  labs(title = "PCA of Mouse WT vs DKO ABC Transcriptomes",
       x = "PC1", y = "PC2", color = "Disease State") +
  scale_color_manual(values=c("DKO" = 'red', "WT" = 'cornflowerblue')) +
  theme_classic() +
  theme(axis.text = element_blank(),
        axis.title = element_text(size = 24),
        axis.ticks = element_blank(), 
        plot.title = element_text(hjust=0.5, size = 24),
        legend.title = element_text(size = 18),
        legend.text = element_text(size = 18))

#Print PCA to pdf
pdf(paste("../Figures/PCA_mouse",".pdf", sep = ""), 
    width = 10, height = 10) 
print(plt)
dev.off()

#What drives PCA separation for each?
res.var <- get_pca_var(pca) #variance of pca explained by each PC
# PC Contributions summed over first and second PC
top_cont <- sort((abs(res.var$contrib[,1]) + abs(res.var$contrib[,2])),
                 decreasing = TRUE) 
top_cont

#From top_cont, choose top-2 contributors and plot **DOESNT REALLY WORK, issue w using EntrezID to pull data
p9 <- ggplot(as.data.frame(hX), aes(x=DCK, y = TNFAIP8)) +
  geom_point(alpha = .35, size = 5, 
             aes(fill = hID, color = hID)) +
  labs(title = "Contribution of Top-2 Features to Diagnostic Separation") +
  scale_color_manual(values=c("SLE" = 'red', "Control" = 'cornflowerblue')) +
  theme_classic() +
  theme(axis.text = element_text(size = 20),
        axis.title = element_text(size = 24),
        axis.ticks = element_blank(), 
        plot.title = element_text(hjust=0.5, size = 24),
        legend.title = element_text(size = 18),
        legend.text = element_text(size = 18))
p9

```

# Differential Expression Analysis ("DE")

```{r Differential Expression}

#Create identification matrix designating each sample as DKO or WT
matrix_mouse <- model.matrix(~0+ID)
matrix_human <- model.matrix(~0+hID)
colnames(matrix_mouse) <- c("DKO","WT")
colnames(matrix_human) <- c("Control","SLE")

#Fit a linear model of gene counts by the design matrix, 
#so each group is compared. Transpose cpm counts to align with design matrix.
fit_human <- lmFit(t(hcounts), matrix_human)
fit_mouse <- lmFit(t(counts), matrix_mouse)

#makeContrasts takes expression specifying contrasts, returns contrast matrix
contrasts_mouse <- makeContrasts(DKO - WT, levels=matrix_mouse)
contrasts_human <- makeContrasts(SLE - Control, levels=matrix_human)
#Compute estimated coefficients from linear model fit, contrasts
fit2_mouse <- contrasts.fit(fit_mouse, contrasts_mouse)
fit2_human <- contrasts.fit(fit_human, contrasts_human)

#Takes microarray linear fit model, returns t-statistics, F-statistics, and 
#log-odds of differential expression using empirical Bayes' moderation
fit2_mouse <- eBayes(fit2_mouse)
fit2_human <- eBayes(fit2_human)

#Check how many genes are significantly over-expressed or under-expressed
#between conditions, and which are equally expressed (~10% of genes are 
#significantly diff. expressed)
table(decideTests(fit2_mouse)) #returns # of over (+1) or under (-1) expressed genes
table(decideTests(fit2_human))
#Get table of top-ranked genes, make gene names explicit variable
full_results_mouse <- topTable(fit2_mouse, number=Inf, adjust.method="BH")
full_results_mouse <- tibble::rownames_to_column(full_results_mouse,"ID")
full_results_human <- topTable(fit2_human, number=Inf, adjust.method="BH")
full_results_human <- tibble::rownames_to_column(full_results_human,"ID")

p_cutoff <- 0.05
fc_cutoff <- 1
topN <- 100 #top N genes for annotation

#Reformat full results to have additional variables of interest, filter for 
#set parameters, plot volcano plot of DE and significance
full_results_human <- full_results_human %>% 
  mutate(Significant = adj.P.Val < p_cutoff, abs(logFC) > fc_cutoff ) %>% 
  mutate(Rank = 1:n(), Label = ifelse(Rank < topN, ID,"")) %>%
  mutate(logpval = -log10(adj.P.Val))
full_results_mouse <- full_results_mouse %>% 
  mutate(Significant = adj.P.Val < p_cutoff, abs(logFC) > fc_cutoff ) %>% 
  mutate(Rank = 1:n(), Label = ifelse(Rank < topN, ID,"")) %>%
  mutate(logpval = -log10(adj.P.Val))

#Plot human or mouse volcano plots by switching title, full_results_(desired species)
plt <- ggplot(full_results_human, aes(x = logFC, y = logpval, col=Significant,
                                      label=Label)) + 
  geom_point() + 
  geom_text_repel(col="black", max.overlaps = 10) +
  ggtitle("Differential Expression Analysis- Human Control vs SLE B Cells") +#WT vs DKO ABCs") + #
  xlab("Log fold-change") + ylab("-log(p-value)") +
  xlim(-50,50) + #ylim(-5, -1) +
  theme_light() +
  theme(plot.title = element_text(hjust=0.5, size = 18))  

#Print volano to pdf
pdf(paste("../Figures/Volcano_Human",".pdf", sep = ""), 
    width = 10, height = 8) 
print(plt)
dev.off()

```
#Which genes are differentially expressed in both mouse and human models of SLE?

```{r DE overlaps}

#Which pathways and genes are differentially expressed in both the mouse and human models?

#Make sure they are in the same direction! ie control > or < SLE in BOTH
#Both must also be significantly differentially expressed
higher_de_mouse <- subset(full_results_mouse, logFC > 0 & Significant == TRUE)
lower_de_mouse <- subset(full_results_mouse, logFC < 0 & Significant == TRUE)
higher_de_human <- subset(full_results_human, logFC > 0 & Significant == TRUE)
lower_de_human <- subset(full_results_human, logFC < 0 & Significant == TRUE)

high_de_overlap <- higher_de_human[which(toupper(higher_de_mouse$ID) %in% higher_de_human$ID),]
low_de_overlap <- lower_de_human[which(toupper(lower_de_mouse$ID) %in% lower_de_human$ID),]
de_overlap <- rbind(high_de_overlap, low_de_overlap)

#Use Entrez IDs (column 2) from overlap pathways to identify GO and KEGG enriched pathways
human_gene_IDs <- mapIds(org.Hs.eg.db, de_overlap$ID, 'ENTREZID', 'SYMBOL')
de_overlap$EntrezID <- human_gene_IDs

# Convert GO terms in publicly available db to a list
human_GO_terms <- org.Hs.egGO
# Get the entrez gene identifiers that are mapped to a GO ID
mapped_genes <- mappedkeys(human_GO_terms)
# Convert to a list
human_GO_list <- as.list(human_GO_terms[mapped_genes])


```


# Create heatmap for 2 page report

```{r Heatmap}

#Temporarily only use top-100 most significant differentially expressed genes
Xtmp <- X[,1:100]

#Cluster genes using hierarchical clustering with average Euclidean distance
order_cluster_cols <- hclust(dist(t(Xtmp)), method="average")

#Define sorting function to sort dendrogram/clustering distances by magnitude
sort_hclust <- function(...) as.hclust(dendsort(as.dendrogram(...)))

#Sort columns and rows according to sorting function
mat_cluster_cols <- sort_hclust(order_cluster_cols)
mat_cluster_rows <- sort_hclust(hclust(dist(X)))

#Define quantile_breaks function to take given matrix, split into 10 equal
#pieces, and return break locations. This will be used to color the heatmap
#with each color representing 10% of the data, to distinguish deciles
quantile_breaks <- function(xs, n = 10) {
  breaks <- quantile(xs, probs = seq(0, 1, length.out = n), na.rm = TRUE)
  breaks[!duplicated(breaks)]
}

#Use break function to create breaks in gene expression data 
mat_breaks <- quantile_breaks(Xtmp, n = 10)

#Label heatmap rows with sample type, columns with gene names
annotation_row = data.frame(strain = ID)
rownames(annotation_row) <- rownames(Xtmp)

#Provide colors for sample types
my_colors <- list(
  strain = c("WT" = '#8C02CC', "DKO" = '#5192C3'))

#generate heatmap of log2 cpm count for top-100 differentially expressed genes,
#using sample color assignment, clustering columns and rows in previously sorted 
#order from sort_hclust, breaks from previously defined mat_breaks,
#and providing specific font,cell, and dedrogram (tree) sizes

plt <- pheatmap(Xtmp, color = viridis(9), 
         annotation_colors = my_colors, 
         annotation_row = annotation_row, 
         cellwidth = 8, 
         cellheight = 18, 
         fontsize = 12, 
         border_color = "grey", 
         treeheight_row = 3, 
         treeheight_col = 12,
         breaks = mat_breaks,
         cluster_cols = mat_cluster_cols,
         cluster_rows = mat_cluster_rows,
         cex = 0.7
         )

#Print heatmap to pdf
pdf(paste("../Figures/Pheatmap",".pdf", sep = ""), 
    width = 20, height = 8) 
print(plt)
dev.off()
```

#Differential connectivity

```{r Package loading}

library(org.Hs.eg.db)

#BiocManager::install("org.Mm.eg.db")

library(org.Mm.eg.db)
library(githubinstall)
#githubinstall("dnapath")
library(dnapath)
#BiocManager::install("reactome.db")
#BiocManager::install("topGO")
library(reactome.db)
library(SeqNet)
library(topGO)
library(tidyverse)
```

``` {r DC analysis}
#Convert gene names to Entrez IDs
mouse_gene_names <- mapIds(org.Mm.eg.db, colnames(counts), 'ENTREZID', 'SYMBOL')
human_gene_names <- mapIds(org.Hs.eg.db, colnames(hcounts), 'ENTREZID', 'SYMBOL')

#Add entrez IDs to dataframe
mouse_Entrez_genes <- data.frame(counts)
colnames(mouse_Entrez_genes) <- mouse_gene_names
human_Entrez_genes <- data.frame(hcounts)
colnames(human_Entrez_genes) <- human_gene_names

#Get reactome pathways represented within genes measured for both mouse & human
react_paths_human <- get_reactome_pathways(species = "Homo sapiens")
react_paths_mouse <- get_reactome_pathways(species = "Mus musculus")

# Run dnapath using gene expression and group information for human data
results_human <- dnapath(human_Entrez_genes, 
                   pathway_list = react_paths_human, 
                   groups = df_diagnosis$diagnosis,
                   network_inference = run_pcor)

results_mouse <- dnapath(mouse_Entrez_genes, 
                   pathway_list = react_paths_mouse, 
                   groups = ID,
                   network_inference = run_corr)

results_human <- filter_pathways(results_human, alpha_pathway = 0.01)
results_mouse <- filter_pathways(results_mouse,alpha_pathway = 0.333) #limited by replicate number
# Setting seed allows for reproducible work
set.seed(123)
results_mouse <- rename_genes(results_mouse, to = "symbol", species = "Mus musculus",
                        dir_save = tempdir())
results_human <- rename_genes(results_human, to = "symbol", species = "Homo Sapiens",
                        dir_save = tempdir())

plot(results_human[[100]])

# Extract the two estimated association networks for the first pathway
nw <- get_networks(results_human[[100]])
# Plot the networks using the SeqNet::plot_network function.

# Plot the two networks (in separate plots)
g <- SeqNet::plot_network(nw[[1]])
SeqNet::plot_network(nw[[1]], compare_graph = g)
# Plot of the differential network for pathway 1.
# Use compare_graph to maintain same layout
plot(results_human[[1]], compare_graph = g)

#Get dnapath_list format into dataframe for both mouse and human
pathway <- list()
edges <- list()
dc_score <- list()
pval <- list()
nw1 <- list()
nw2 <- list()

for (path in 1:length(results_human))
{
  tmp <- summary(results_human[[path]], by_gene = FALSE)
  pathway <- append(pathway, tmp$pathway)
  edges <- append(edges, tmp$edges)
  dc_score <- append(dc_score, tmp$dc_score)
  pval <- append(pval, tmp$p_value)
  nw1 <- append(nw1, tmp$nw1)
  nw2 <- append(nw2, tmp$nw2)
}

df_human <- data.frame(pathway = unlist(pathway),
                       edges = unlist(edges),
                       dc_score = unlist(dc_score),
                       pval = unlist(pval),
                       nw1 = unlist(nw1),
                       nw2 = unlist(nw2))

pathway <- list()
edges <- list()
dc_score <- list()
pval <- list()
nw1 <- list()
nw2 <- list()

for (path in 1:length(results_mouse))
{
  tmp <- summary(results_mouse[[path]], by_gene = FALSE)
  pathway <- append(pathway, tmp$pathway)
  edges <- append(toupper(edges), tmp$edges)
  dc_score <- append(dc_score, tmp$dc_score)
  pval <- append(pval, tmp$p_value)
  nw1 <- append(nw1, tmp$nw1)
  nw2 <- append(nw2, tmp$nw2)
}

df_mouse <- data.frame(pathway = unlist(pathway),
                       edges = unlist(edges),
                       dc_score = unlist(dc_score),
                       pval = unlist(pval),
                       nw1 = unlist(nw1),
                       nw2 = unlist(nw2))

#Which pathways are differentially connected in both the mouse and human models?

#Make sure they are in the same direction! ie control > or < SLE in BOTH

higher_dc_mouse <- subset(df_mouse, nw1 > nw2)
lower_dc_mouse <- subset(df_mouse, nw1 < nw2)
higher_dc_human <- subset(df_human, nw1 > nw2)
lower_dc_human <- subset(df_human, nw1 < nw2)

#Pull only differentially expressed genes out for each
#higher_dc_mouse <- subset(higher_dc_mouse, pval < 0.35)
#lower_dc_mouse <- subset(lower_dc_mouse, pval < 0.35)
#higher_dc_human <- subset(higher_dc_human, pval < 0.05)
#lower_dc_human <- subset(lower_dc_human, pval < 0.05)

high_overlap <- higher_dc_human[which(higher_dc_human$edges %in% higher_dc_mouse$edges),]
low_overlap <- lower_dc_human[which(lower_dc_human$edges %in% lower_dc_mouse$edges),]
low_overlap$dc_score <- -(low_overlap$dc_score)
human_overlap <- rbind(high_overlap, low_overlap)

high_overlap <- higher_dc_mouse[which(higher_dc_mouse$edges %in% higher_dc_human$edges),]
low_overlap <- lower_dc_mouse[which(lower_dc_mouse$edges %in% lower_dc_human$edges),]
low_overlap$dc_score <- -(low_overlap$dc_score)
mouse_overlap <- rbind(high_overlap, low_overlap)


#Use Entrez IDs (column 2) from overlap pathways to identify GO and KEGG enriched pathways
```

#DC Volcano plots

```{r DC Volcano plots}
#to plot human or mouse plots, change ggplot input human_overlap <-> mouse_overlap, and title
p_cutoff <- 0.34
dc_cutoff <- 0.1
topN <- 20 #top N genes for annotation

#Reformat full results to have additional variables of interest, filter for 
#set parameters, plot volcano plot of DE and significance
human_overlap <- human_overlap %>% 
  mutate(Significant = pval < p_cutoff, abs(dc_score) > dc_cutoff ) %>% 
  mutate(Rank = 1:n(), Label = ifelse(Rank < topN, edges,"")) %>%
  mutate(logpval = -log10(pval))
mouse_overlap <- mouse_overlap %>% 
  mutate(Significant = pval < p_cutoff, abs(dc_score) > dc_cutoff ) %>% 
  mutate(Rank = 1:n(), Label = ifelse(Rank < topN, edges,"")) %>%
  mutate(logpval = -log10(pval))

#Plot human or mouse volcano plots by switching title, full_results_(desired species)
plt <- ggplot(mouse_overlap, aes(x = dc_score, y = logpval, col=Significant,
                                      label=Label)) + 
  geom_jitter(alpha = 0.35) + #change to geom_point for human
  geom_text_repel(col="black", max.overlaps = 10) +
  ggtitle("Differential Connectivity Analysis- WT vs DKO ABCs") + #Human Control vs SLE B Cells") +#
  xlab("Differential Connectivity Score") + ylab("-log(p-value)") +
  ylim(-.1,1) + #ylim(-5, -1) +
  theme_light() +
  theme(plot.title = element_text(hjust=0.5, size = 18))  

#Print volano to pdf
pdf(paste("../Figures/Volcano_DC_Mouse",".pdf", sep = ""), 
    width = 10, height = 8) 
print(plt)
dev.off()


```


#Use complex heatmap to overlay 4 heatmaps: 1) mouse DE genes, 2) human DE genes, 3) mouse DC genes (2 row, use mean expr 1 and mean expr 2 z-scored by column), 4) human DC genes (same as mouse)

```{r Complex Heatmap}

```

