---
title: "Differential Expression Analysis"
author: "Bri Ko"
date: "4/11/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# Load packages

```{r load, include=FALSE}
library(readr)
library(dplyr) # for %<% mutating
library(ggplot2) # for plotting
library(ggrepel) # for plot text spacing
library(Rgraphviz) # for visualizing GO graphs
library(gridExtra) # for printing table as pdf
library(limma)
library(edgeR)
library(topGO)
set.seed(1234)  # random seed for reproducibility
```
# Functions

```{r Useful functions}
empiricalBayes <- function(cnts, ID, contrast, p_cutoff, fc_cutoff, topN) {
  ### Input:
  #     cnts: count data (samples x genes)
  #     ID: list of sample ptypes
  #     contrast: string of contrast formula (e.g. "DKO-WT" or "SLE-HC")
  #     p_cutoff: p-value threshold for significance
  #     fc_cutoff: abs(log fold change) threshold for significance
  #     topN: number of top genes for annotation
  
  ### Output:
  #     res: DE analysis results 
  
  # Create identification matrix mapping samples with ptype
  mat <- model.matrix(~0+ID)
  colnames(mat) <- substr(colnames(mat),3,nchar(colnames(mat)))
  
  # Fit a linear model of gene counts by the design matrix, so each group is compared. 
  # Transpose cpm counts to align with design matrix.
  model <- lmFit(t(cnts), mat)
  
  # Create contrast matrix
  cont_mat <- makeContrasts(contrasts=contrast, levels=mat)
  
  # Compute estimated coefficients from linear model fit, contrasts
  model2 <- contrasts.fit(model, cont_mat)
  
  # Takes microarray linear fit model, returns t-statistics, F-statistics, and 
  # log-odds of differential expression using empirical Bayes' moderation
  model2 <- eBayes(model2)
  
  # Check how many genes are significantly over-expressed or under-expressed
  # between conditions, and which are equally expressed
  table(decideTests(model2)) #returns # of over (+1) or under (-1) expressed genes
  
  # Get table of top-ranked genes, make gene names explicit variable
  res <- topTable(model2, number=Inf, adjust.method="BH")
  res <- tibble::rownames_to_column(res,"ID")
  
  #Reformat full results to have additional variables of interest, filter for 
  #set parameters, plot volcano plot of DE and significance
  res <- res %>% 
    mutate(Significant = adj.P.Val < p_cutoff, abs(logFC) > fc_cutoff ) %>% 
    mutate(Rank = 1:n(), Label = ifelse(Rank < topN, ID,"")) %>%
    mutate(logpval = -log10(adj.P.Val))

  return(res)
}

GOenrich <- function(geneList, genes, topGO_args) {
  ### Input
  #     geneList: DE analysis result values for enrichment analysis (e.g. pval)
  #     genes: list of genes of type specified in ID
  #     topGO_args: arguments for new topGOdata object
  #       - ontology: what kind of ontology to use (e.g. BP)
  #       - geneSelectFun: function to select significant genes??? INCORPORATE PVAL<0.05 HERE?
  #       - mapping: string of what database to use for mapping
  #       - ID: type of gene ID (e.g. symbol, entrez)

  ### Output
  #     ???: GO enriched terms and stats

  names(geneList) <- genes

  # Create topGOdata object
  GOdata <- new("topGOdata",
                ontology = go_args[[1]],
                allGenes = geneList,
                geneSelectionFun = go_args[[2]],
                annot = annFUN.org,
                mapping = go_args[[3]],
                ID = go_args[[4]])

  # Run several kinds of enrichment tests
  resFisher <- runTest(GOdata, algorithm = "classic", statistic = "fisher")
  resKS <- runTest(GOdata, algorithm = "classic", statistic = "ks")
  resKS.elim <- runTest(GOdata, algorithm = "elim", statistic = "ks")

  # Generate table of results of enrichment results
  allRes <- GenTable(GOdata, classicFisher = resFisher,
                     classicKS = resKS, elimKS = resKS.elim,
                     orderBy = "elimKS", ranksOf = "classicFisher",
                     topNodes = 20)

  res <- list(GOdata, resFisher, resKS, resKS.elim, allRes)
  return(res)
}
```

# Download and clean data

```{r GEO data from downloaded csv}
# same as PCA_heatmap.Rmd code w/o normalized counts (e.g. mX, hX)
### Mouse data
# Import the data and pull out gene names, counts (log2 cpm)
mdata <- read.csv(file = '../Data/ABC_WT_DKO.csv')
mgenes <- mdata[,"genes"]
mgenes <- mgenes[complete.cases(mgenes)]
cpm_cols <- c("WT_1_log2.cpm.","WT_2_log2.cpm.", 
              "DKO_1_log2.cpm.","DKO_2_log2.cpm.")
mcounts <- mdata[mdata$genes %in% mgenes, cpm_cols]
rownames(mcounts) <- mgenes
mcounts <- 2^mcounts # retrieve raw cpm values
colnames(mcounts) <- substr(colnames(mcounts), 1, nchar(colnames(mcounts))-nchar("_log2.cpm."))
# write.table(mcounts, file.path("../Data/mcounts.txt"), quote=FALSE, sep = "\t", dec = ".",
#             row.names = TRUE, col.names = TRUE)
mcounts <- t(mcounts) # samples x genes
mID <- sub("\\_.*", "", rownames(mcounts)) # list of sample ptypes

### Repeat for human data (in fpkm! raw counts = fpkm*transcipt_length...)
hdata <- read.csv(file = '../Data/Human_SLE_Ctrl.csv')
hgenes <- hdata[,c("ENTREZID","SYMBOL")]
hgenes <- hgenes[complete.cases(hgenes),]
hcounts <- hdata[hdata$ENTREZID %in% hgenes$ENTREZID,4:ncol(hdata)]
hcounts <- hcounts[-which(grepl("2-Mar",hgenes[,2])),]
hgenes <- hgenes[-which(grepl("2-Mar",hgenes[,2])),]
rownames(hcounts) <- hgenes[,2]
colnames(hcounts) <- substr(colnames(hcounts), 1, nchar(colnames(hcounts))-nchar(".fpkm"))
# write.table(hcounts, "hcounts.txt", quote=FALSE, sep = "\t", dec = ".",
#             row.names = TRUE, col.names = TRUE)
hcounts <- t(hcounts)
hID <- sub("\\..*", "", rownames(hcounts))

```

# Differential Expression Analysis ("DE")

```{r Differential Expression}
### Mouse
mres <- empiricalBayes(mcounts, mID, contrast="DKO-WT", 
                       p_cutoff=0.05, fc_cutoff=1, topN=0.2*dim(mcounts)[2])

### Human
hres <- empiricalBayes(hcounts, hID, contrast="SLE-HC", 
                       p_cutoff=0.05, fc_cutoff=1, topN=0.2*dim(hcounts)[2])

```
# DEG overlap between human and mouse

```{r DE overlaps}
# Map human ENTREZ ids -> Symbol
h_entrez <- mapIds(org.Hs.eg.db, keys=hres$ID, column="SYMBOL", keytype="ENTREZID")
hres$Symbol <- h_entrez

# Identify up- and down-regulated genes in both species
m_updegs <- subset(mres, logFC > 0 & Significant == TRUE)
m_downdegs <- subset(mres, logFC < 0 & Significant == TRUE)
h_updegs <- subset(hres, logFC > 0 & Significant == TRUE)
h_downdegs <- subset(hres, logFC < 0 & Significant == TRUE)

# Identify overlapping genes in both species
up_overlap <- intersect(toupper(m_updegs$ID), h_updegs$Symbol)
down_overlap <- intersect(toupper(m_downdegs$ID), h_downdegs$Symbol)
overlap <- c(up_overlap, down_overlap)

# Convert GO terms in publicly available db to a list?
human_GO_terms <- org.Hs.egGO
mapped_genes <- mappedkeys(human_GO_terms) # entrez:GO mapping
human_GO_list <- as.list(human_GO_terms[mapped_genes])
# GET OVERLAPPING GENES' ASSOCIATED GO TERMS?

```
# GO Enrichment of DEGs using topGO

``` {r GO enrichment DEGs}
# GO enrichment of DEGs
# documentation: https://www.bioconductor.org/packages/devel/bioc/vignettes/topGO/inst/doc/topGO.pdf
# walkthru: https://ucdavis-bioinformatics-training.github.io/2018-June-RNA-Seq-Workshop/friday/enrichment.html
### Mouse
# Set up input arguments
# geneSelectFun <- function(x)
geneSelectFun <- function(x){ return(x < 0.05)}
go_args <- c("BP", geneSelectFun, "org.Mm.eg.db", "SYMBOL")

# Output GO enrichment results
mGOres <- GOenrich(mres$adj.P.Val, mres$ID, go_args) 
mGOdata <- mGOres[[1]]
mresFisher <- mGOres[[2]]
mresKS <- mGOres[[3]]
mresKS.elim <- mGOres[[4]]
mallRes <- mGOres[[5]]
showSigOfNodes(mGOdata, score(mresKS.elim), firstSigNodes = 5, useInfo = 'all')

m_nSigTerms <- sum(score(mresKS.elim)<0.05)

### Human
# Set up input arguments
go_args <- c("BP", geneSelectFun, "org.Hs.eg.db", "ENTREZ")

# Output GO enrichment results as c(GOdata, resFisher, resKS, resKS.elim, allRes)
hGOres <- GOenrich(hres$adj.P.Val, hres$ID, go_args) 
hGOdata <- hGOres[[1]]
hresFisher <- hGOres[[2]]
hresKS <- hGOres[[3]]
hresKS.elim <- hGOres[[4]]
hallRes <- hGOres[[5]]
showSigOfNodes(hGOdata, score(hresKS.elim), firstSigNodes = 5, useInfo = 'all')

h_nSigTerms <- sum(score(hresKS.elim)<0.05)
```
# GO Enrichment of overlapping genes using topGO

``` {r GO enrichment overlap}
# Set up input arguments and subset human data for overlap
# geneSelectFun <- function(x)x
selection <- function(allScore){ return(allScore < 0.05)}
go_args <- c("BP", geneSelectFun, "org.Hs.eg.db", "SYMBOL")
hres_overlap <- hres[which(hres$Symbol %in% overlap),]

# Output GO enrichment results
GOres <- GOenrich(hres_overlap$adj.P.Val, hres_overlap$Symbol, go_args) 
GOdata <- GOres[[1]]
resFisher <- GOres[[2]]
resKS <- GOres[[3]]
resKS.elim <- GOres[[4]]
allRes <- GOres[[5]]
showSigOfNodes(GOdata, score(resKS.elim), firstSigNodes = 5, useInfo = 'all')

nSigTerms <- sum(score(resKS.elim)<0.05)

```
# GSEA stuff

```{r GSEA}
### For GSEA Pre-ranked
mprernk <- data.frame(name = mres$ID,
                      metric = sign(mres$logFC) * mres$logpval)
hprernk <- data.frame(name = hres$ID,
                      metric = sign(hres$logFC) * hres$logpval)

# write.table(mprernk, file=file.path("../Data/mprernk_deg.rnk"),
#             row.names=FALSE, sep="\t", quote=FALSE)
# write.table(hprernk, file=file.path("../Data/hprernk_deg.rnk"),
#             row.names=FALSE, sep="\t", quote=FALSE)

### Mouse and human non-SLE gene set overlap
hctl_file <- file.path("../GSEA/human_DEGprerank/gsea_report_for_na_neg_1620851989820.tsv")
mctl_file <- file.path("../GSEA/mouse_DEGprerank/gsea_report_for_na_neg_1620853009801.tsv")
hgsea_ctl <- read.delim(hctl_file) # 584 sets
mgsea_ctl <- read.delim(mctl_file) # 396 sets

hgsea_ctl05 <- hgsea_ctl[which(hgsea_ctl$FDR.q.val<0.25 & hgsea_ctl$NOM.p.val<0.05),
                         c("NAME","SIZE","NES","NOM.p.val","FDR.q.val")]
mgsea_ctl05 <- mgsea_ctl[which(mgsea_ctl$FDR.q.val<0.25 & mgsea_ctl$NOM.p.val<0.05),
                         c("NAME","SIZE","NES","NOM.p.val","FDR.q.val")]
ctl_overlap <- intersect(hgsea_ctl05$NAME, mgsea_ctl05$NAME)

### Mouse and human SLE gene set overlap
hsle_file <- file.path("../GSEA/human_DEGprerank/gsea_report_for_na_pos_1620851989820.tsv")
msle_file <- file.path("../GSEA/mouse_DEGprerank/gsea_report_for_na_pos_1620853009801.tsv")
hgsea_sle <- read.delim(hsle_file) # 215 sets
mgsea_sle <- read.delim(msle_file) # 395 sets

hgsea_sle05 <- hgsea_sle[which(hgsea_sle$FDR.q.val<0.25 & hgsea_sle$NOM.p.val<0.05),
                         c("NAME","SIZE","NES","NOM.p.val","FDR.q.val")]
mgsea_sle05 <- mgsea_sle[which(mgsea_sle$FDR.q.val<0.25 & mgsea_sle$NOM.p.val<0.05),
                         c("NAME","SIZE","NES","NOM.p.val","FDR.q.val")]
sle_overlap <- intersect(hgsea_sle05$NAME, mgsea_sle05$NAME)

### Dotplots
# No SLE overlap; ignore this condition
ggplot(hgsea_ctl05[hgsea_ctl05$NAME %in% ctl_overlap,], 
       aes(x=-log10(NOM.p.val),
           y=substring(NAME,10,50),
           size=SIZE,
           color=NES)) +
  geom_point() +
  scale_color_gradient(low = "blue", high = "red") +
  theme_bw() +
  ylab("Reactome Pathways") +
  xlab("-log10(pval)") +
  ggtitle("Overlapping DEG Pathways - Human HC")
ggplot(mgsea_ctl05[mgsea_ctl05$NAME %in% ctl_overlap,], 
       aes(x=-log10(NOM.p.val),
           y=substring(NAME,10,50),
           size=SIZE,
           color=NES)) +
  geom_point() +
  scale_color_gradient(low = "blue", high = "red") +
  theme_bw() +
  ylab("Reactome Pathways") +
  xlab("-log10(pval)") +
  ggtitle("Overlapping DEG Pathways - Mouse WT")


```
# Visualization

```{r visualize functional enrichment}
### Mouse dotplot
df_dot <- mallRes[1:20,c("Term","elimKS","Significant")]
df_dot$elimKS <- as.numeric(df_dot$elimKS)
colnames(df_dot)[3] <- "nSigGenes"
ggplot(df_dot, aes(x=-log10(elimKS), y=reorder(Term, -elimKS), color=nSigGenes)) + 
  geom_point(size=4) +
  scale_color_gradient(low = "blue", high = "red") +
  theme_bw() + 
  ylab("GO Terms") + 
  xlab("-log(KS elim pval)") + 
  ggtitle("Top 20 Most Significant GO Terms in Mouse")

### Human dotplot
df_dot <- hallRes[1:20,c("Term","elimKS","Significant")]
df_dot$elimKS <- as.numeric(df_dot$elimKS)
colnames(df_dot)[3] <- "nSigGenes"
ggplot(df_dot, aes(x=-log10(elimKS), y=reorder(Term,-elimKS), color=nSigGenes)) + 
  geom_point(size=4) +
  scale_color_gradient(low = "blue", high = "red") +
  theme_bw() + 
  ylab("GO Terms") + 
  xlab("-log(KS elim pval)") + 
  ggtitle("Top 20 Most Significant GO Terms in Human")

### Overlap dotplot
df_dot <- allRes[1:20,c("Term","elimKS","Significant")]
df_dot$elimKS <- as.numeric(df_dot$elimKS)
colnames(df_dot)[3] <- "nSigGenes"
ggplot(df_dot, aes(x=-log10(elimKS), y=reorder(Term,-elimKS), color=nSigGenes)) + 
  geom_point(size=4) +
  scale_color_gradient(low = "blue", high = "red") +
  theme_bw() + 
  ylab("GO Terms") + 
  xlab("-log(KS elim pval)") + 
  ggtitle("Top 20 Most Significant Species-Overlapping GO Terms")

# Save dot plots to pdf
# had to manually save
```