---
title: "DC Network Analysis"
author: "Bri Ko"
date: "4/12/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# Load packages

```{r load, include=FALSE}
library(readr)
library(dplyr) # for %<% mutating
library(ggplot2) # for plotting
library(ggrepel) # for plot text spacing
library(gridExtra) # for printing table as pdf
library(tidyverse)
library(limma)
library(dnapath)
library(topGO)
library(SeqNet)
library(org.Hs.eg.db)
library(org.Mm.eg.db)
library(reactome.db)
set.seed(1234)  # random seed for reproducibility
```
# Functions

```{r Useful functions}
GOenrich <- function(geneList, genes, topGO_args) {
  ### Input
  #     geneList: DE analysis result values for enrichment analysis (e.g. pval)
  #     genes: list of genes of type specified in ID
  #     topGO_args: arguments for new topGOdata object
  #       - ontology: what kind of ontology to use (e.g. BP)
  #       - geneSelectFun: function to select significant genes??? INCORPORATE PVAL<0.05 HERE?
  #       - mapping: string of what database to use for mapping
  #       - ID: type of gene ID (e.g. symbol, entrez)
  
  ### Output
  #     ???: GO enriched terms and stats
  
  names(geneList) <- genes
  
  # Create topGOdata object
  GOdata <- new("topGOdata",
                ontology = go_args[[1]],
                allGenes = geneList,
                geneSelectionFun = go_args[[2]],
                annot = annFUN.org, 
                mapping = go_args[[3]],
                ID = go_args[[4]])
  
  # Run several kinds of enrichment tests
  resFisher <- runTest(GOdata, algorithm = "classic", statistic = "fisher")
  resKS <- runTest(GOdata, algorithm = "classic", statistic = "ks")
  resKS.elim <- runTest(GOdata, algorithm = "elim", statistic = "ks")
  
  # Generate table of results of enrichment results
  allRes <- GenTable(GOdata, classicFisher = resFisher,
                     classicKS = resKS, elimKS = resKS.elim,
                     orderBy = "elimKS", ranksOf = "classicFisher", 
                     topNodes = 20)
  
  res <- list(GOdata, resFisher, resKS, resKS.elim, allRes)
  return(res)
}
```
# Download and clean data

```{r GEO data from downloaded csv}
# same as PCA_heatmap.Rmd code w/o normalized counts (e.g. mX, hX)
### Mouse data
# Import the data and pull out gene names, counts (log2 cpm)
mdata <- read.csv(file = '../Data/ABC_WT_DKO.csv')
mgenes <- mdata[,"genes"]
mgenes <- mgenes[complete.cases(mgenes)]
cpm_cols <- c("WT_1_log2.cpm.","WT_2_log2.cpm.", 
              "DKO_1_log2.cpm.","DKO_2_log2.cpm.")
mcounts <- mdata[mdata$genes %in% mgenes, cpm_cols]
rownames(mcounts) <- mgenes
mcounts <- t(mcounts) # samples x genes
mcounts <- 2^mcounts # retrieve raw cpm values
mID <- sub("\\_.*", "", rownames(mcounts)) # list of sample ptypes

### Repeat for human data (in fpkm! raw counts = fpkm*transcipt_length...)
hdata <- read.csv(file = '../Data/Human_SLE_Ctrl.csv')
hgenes <- hdata[,c("ENTREZID","SYMBOL")]
hgenes <- hgenes[complete.cases(hgenes),]
hcounts <- hdata[hdata$ENTREZID %in% hgenes$ENTREZID,4:ncol(hdata)]
hcounts <- hcounts[-which(grepl("2-Mar",hgenes[,2])),]
hgenes <- hgenes[-which(grepl("2-Mar",hgenes[,2])),]
rownames(hcounts) <- hgenes[,1]
hcounts <- t(hcounts)
hID <- sub("\\..*", "", rownames(hcounts))
```

# Differential connectivity

``` {r DC analysis}
# Convert gene names to Entrez IDs, add to counts matrix
m_entrez <- mapIds(org.Mm.eg.db, keys=colnames(mcounts), column="ENTREZID", keytype="SYMBOL")
h_entrez <- mapIds(org.Hs.eg.db, keys=colnames(hcounts), column="SYMBOL", keytype="ENTREZID")

mouse_Entrez_genes <- data.frame(mcounts)
colnames(mouse_Entrez_genes) <- m_entrez

# Get reactome pathways represented within genes measured for both mouse & human
react_paths_human <- get_reactome_pathways(species = "Homo sapiens")
react_paths_mouse <- get_reactome_pathways(species = "Mus musculus")

# Run dnapath using gene expression and group information for human data
results_human <- dnapath(hcounts, 
                         pathway_list = react_paths_human, 
                         groups = hID,
                         network_inference = run_pcor) # 797 pathways with multiple genes/pathway

results_mouse <- dnapath(mouse_Entrez_genes, 
                         pathway_list = react_paths_mouse, 
                         groups = mID,
                         network_inference = run_corr) # 587 pathways with multiple genes/pathway

hres_sigpath <- filter_pathways(results_human, alpha_pathway = 0.01)
mres_sigpath <- filter_pathways(results_mouse, alpha_pathway = 0.34) #limited by replicate number

# Setting seed allows for reproducible work
set.seed(123)
mres_sigpath <- rename_genes(mres_sigpath, to = "symbol", species = "Mus musculus",
                             dir_save = tempdir())
hres_sigpath <- rename_genes(hres_sigpath, to = "symbol", species = "Homo Sapiens",
                             dir_save = tempdir())

plot(hres_sigpath[[length(hres_sigpath)]])
plot(mres_sigpath[[length(mres_sigpath)]])

# Extract the two estimated association networks for the first pathway
h_nw <- get_networks(hres_sigpath[[length(hres_sigpath)]])
m_nw <- get_networks(mres_sigpath[[length(mres_sigpath)]])

# Plot the networks using the SeqNet::plot_network function.
# h_nwGraph <- SeqNet::plot_network(h_nw[[1]])
# SeqNet::plot_network(h_nw[[1]], compare_graph = h_nwGraph)
# 
# m_nwGraph <- SeqNet::plot_network(m_nw[[1]])
# SeqNet::plot_network(m_nw[[1]], compare_graph = m_nwGraph)

# Plot of the differential network for pathway 1.
# Use compare_graph to maintain same layout
# plot(hres_sigpath[[1]], compare_graph = h_nwGraph) # not working...

#Get dnapath_list format into dataframe for both mouse and human
pathway <- list()
genes <- list()
dc_score <- list()
pval <- list()
mean_expr1 <- list()
mean_expr2 <- list()

for (path in 1:length(results_human))
{
  tmp <- summary(results_human[[path]], by_gene = TRUE)
  pathway <- append(pathway, tmp$pathway)
  genes <- append(genes, tmp$genes)
  dc_score <- append(dc_score, tmp$dc_score)
  pval <- append(pval, tmp$p_value)
  mean_expr1 <- append(mean_expr1, tmp$mean_expr1)
  mean_expr2 <- append(mean_expr2, tmp$mean_expr2)
}

df_human <- data.frame(pathway = unlist(pathway),
                       entrez = unlist(genes),
                       symbol = h_entrez[match(unlist(genes), names(h_entrez))],
                       dc_score = unlist(dc_score),
                       pval = unlist(pval),
                       mean_expr1 = unlist(mean_expr1),
                       mean_expr2 = unlist(mean_expr2))

# for (path in 1:length(hres_sigpath))
# {
#   tmp <- summary(hres_sigpath[[path]], by_gene = TRUE)
#   pathway <- append(pathway, tmp$pathway)
#   genes <- append(genes, tmp$genes)
#   dc_score <- append(dc_score, tmp$dc_score)
#   pval <- append(pval, tmp$p_value)
#   mean_expr1 <- append(mean_expr1, tmp$mean_expr1)
#   mean_expr2 <- append(mean_expr2, tmp$mean_expr2)
# }
# 
# df_human <- data.frame(pathway = unlist(pathway),
#                        genes = unlist(genes),
#                        dc_score = unlist(dc_score),
#                        pval = unlist(pval),
#                        mean_expr1 = unlist(mean_expr1),
#                        mean_expr2 = unlist(mean_expr2))

pathway <- list()
genes <- list()
dc_score <- list()
pval <- list()
mean_expr1 <- list()
mean_expr2 <- list()

for (path in 1:length(results_mouse))
{
  tmp <- summary(results_mouse[[path]], by_gene = TRUE)
  pathway <- append(pathway, tmp$pathway)
  genes <- append(toupper(genes), tmp$genes)
  dc_score <- append(dc_score, tmp$dc_score)
  pval <- append(pval, tmp$p_value)
  mean_expr1 <- append(mean_expr1, tmp$mean_expr1)
  mean_expr2 <- append(mean_expr2, tmp$mean_expr2)
}

df_mouse <- data.frame(pathway = unlist(pathway),
                       entrez = unlist(genes),
                       symbol = names(m_entrez)[match(unlist(genes), m_entrez)],
                       dc_score = unlist(dc_score),
                       pval = unlist(pval),
                       mean_expr1 = unlist(mean_expr1),
                       mean_expr2 = unlist(mean_expr2))

# for (path in 1:length(mres_sigpath))
# {
#   tmp <- summary(mres_sigpath[[path]], by_gene = TRUE)
#   pathway <- append(pathway, tmp$pathway)
#   genes <- append(toupper(genes), tmp$genes)
#   dc_score <- append(dc_score, tmp$dc_score)
#   pval <- append(pval, tmp$p_value)
#   mean_expr1 <- append(mean_expr1, tmp$mean_expr1)
#   mean_expr2 <- append(mean_expr2, tmp$mean_expr2)
# }
# 
# df_mouse <- data.frame(pathway = unlist(pathway),
#                        genes = unlist(genes),
#                        dc_score = unlist(dc_score),
#                        pval = unlist(pval),
#                        mean_expr1 = unlist(mean_expr1),
#                        mean_expr2 = unlist(mean_expr2))

# Overlap in DCGs
df_human_overlap <- unique(df_human$genes[which(df_human$genes %in% df_mouse$genes)])
df_mouse_overlap <- unique(df_mouse$genes[which(df_mouse$genes %in% df_human$genes)])
df_overlap <- data.frame(human = sort(df_human_overlap), mouse = sort(df_mouse_overlap))

#What human pathways are those overlapping genes involved in?
human_overlap <- df_human[which(df_human$genes %in% df_overlap$human),]
human_overlap <- human_overlap[which(human_overlap$pval < 0.05),]
human_overlap <- as.data.frame(count(human_overlap,human_overlap$pathway))
colnames(human_overlap) <- c("Pathway","Count")
human_overlap$Pathway <- sapply(strsplit(human_overlap$Pathway,"See"), `[`, 1)
human_overlap$Pathway <- as.factor(human_overlap$Pathway)


#Plot top 20 pathways (by # of DE genes shared by mouse/human)
#Most of #s15-25 are signalling, so selected these for showing
ordered_overlap <- human_overlap[order(-human_overlap$Count),]
ordered_overlap_a <- ordered_overlap[which(ordered_overlap$Count >5),]
ordered_overlap_b <- ordered_overlap[which(grepl("signal|Caspase|DUB"
                                                 ,ordered_overlap$Pathway)),] #adding to make an even 30
ordered_overlap_b <- ordered_overlap_b[which(ordered_overlap_b$Count == 5),]
oo <- rbind(ordered_overlap_a,ordered_overlap_b)

#Plot of differentially connected networks ordered by # of DC genes in network
p2 <-  ggplot(oo, aes(x=reorder(Pathway, Count), y=Count)) +
    geom_bar(stat="identity", fill="#f68060", alpha=.6) +
    coord_flip() +
    labs(y="Number of Significantly Differentially Connected Genes in Pathway",
         x = "Reactome Pathway",
         title = "Top 30 Human Pathways Most Differentially Connected") +
    theme_classic() +
    theme(axis.text = element_text(size = 40, face = "bold"),
        axis.title = element_text(size = 50),
        axis.ticks = element_blank(), 
        plot.title = element_text(hjust=0.5, size = 50))

#Now only look at Reactome pathways from differentially connected AND expressed
df_human_sort <- data.frame(df_human[order(-df_human$dc_score),])
df_human_sort <- df_human_sort[which(
  abs(log2(df_human_sort$mean_expr1/df_human_sort$mean_expr2)) > 1),]
df_human_sort <- df_human_sort[-which(
  df_human_sort$pval > 0.05),]
path_list <- count(df_human_sort,df_human_sort$pathway)
df_human_sort_rank <- path_list[order(-path_list$n),]
colnames(df_human_sort_rank) <-c("Pathway","Count")
df_human_sort_rank$Pathway <- sapply(strsplit(df_human_sort_rank$Pathway,"See"), `[`, 1)
df_human_sort_rank$Pathway <- as.factor(df_human_sort_rank$Pathway)
  
p3 <-  ggplot(subset(df_human_sort_rank, Count >1),
              aes(x=reorder(Pathway, Count), y=Count)) +
    geom_bar(stat="identity", fill="#f68060", alpha=.6) +
    coord_flip() +
    labs(y="Number of Significantly Differentially Connected   \n and Enriched Genes in Pathway",
         x = "Reactome Pathway",
         title = "Top Human Pathways Most Enriched \n in Differentially Connected and Expressed Genes") +
    theme_classic() +
    theme(axis.text = element_text(size = 40, face = "bold"),
        axis.title = element_text(size = 50),
        axis.ticks = element_blank(), 
        plot.title = element_text(hjust=0.5, size = 50))
```

#DC Volcano plots

```{r DC Volcano plots}
#to plot human or mouse plots, change ggplot input human_overlap <-> mouse_overlap, and title
p_cutoff <- 0.34
dc_cutoff <- 0.1
topN <- 20 #top N genes for annotation

#Reformat full results to have additional variables of interest, filter for 
#set parameters, plot volcano plot of DE and significance
human_overlap <- human_overlap %>% 
  mutate(Significant = pval < p_cutoff, abs(dc_score) > dc_cutoff ) %>% 
  mutate(Rank = 1:n(), Label = ifelse(Rank < topN, genes,"")) %>%
  mutate(logpval = -log10(pval))
mouse_overlap <- mouse_overlap %>% 
  mutate(Significant = pval < p_cutoff, abs(dc_score) > dc_cutoff ) %>% 
  mutate(Rank = 1:n(), Label = ifelse(Rank < topN, genes,"")) %>%
  mutate(logpval = -log10(pval))

#Plot human or mouse volcano plots by switching title, full_results_(desired species)
plt <- ggplot(mouse_overlap, aes(x = dc_score, y = logpval, col=Significant,
                                      label=Label)) + 
  geom_jitter(alpha = 0.35) + #change to geom_point for human
  geom_text_repel(col="black", max.overlaps = 10) +
  ggtitle("Differential Connectivity Analysis- WT vs DKO ABCs") + #Human Control vs SLE B Cells") +#
  xlab("Differential Connectivity Score") + ylab("-log(p-value)") +
  ylim(-.1,1) + #ylim(-5, -1) +
  theme_light() +
  theme(plot.title = element_text(hjust=0.5, size = 18))  

```

# GO Enrichment of overlapping DCs using topGO

``` {r GO enrichment}
# GO enrichment of DEGs
# documentation: https://www.bioconductor.org/packages/devel/bioc/vignettes/topGO/inst/doc/topGO.pdf
# walkthru: https://ucdavis-bioinformatics-training.github.io/2018-June-RNA-Seq-Workshop/friday/enrichment.html
# Set up input arguments
geneSelectFun <- function(x)x
go_args <- c("BP", geneSelectFun, "org.Mm.eg.db", "SYMBOL")

### Mouse
# Convert genes to mouse format
library(stringr)
df_mouse$genes <- str_to_title(df_mouse$genes)

# Output GO enrichment results as c(GOdata, resFisher, resKS, resKS.elim, allRes)
mGOres <- GOenrich(df_mouse$pval, df_mouse$genes, go_args) 
mGOdata <- mGOres[[1]]
mresFisher <- mGOres[[2]]
mresKS <- mGOres[[3]]
mresKS.elim <- mGOres[[4]]
mallRes <- mGOres[[5]]

### Human
go_args <- c("BP", geneSelectFun, "org.Hs.eg.db", "SYMBOL")
hGOres <- GOenrich(df_human$pval, df_human$genes, go_args) 
hGOdata <- hGOres[[1]]
hresFisher <- hGOres[[2]]
hresKS <- hGOres[[3]]
hresKS.elim <- hGOres[[4]]
hallRes <- hGOres[[5]]

### Overlap
overlapDCG <- df_human[which(df_human$genes %in% df_human_overlap),]
GOres <- GOenrich(overlapDCG$pval, overlapDCG$genes, go_args) 
GOdata <- GOres[[1]]
resFisher <- GOres[[2]]
resKS <- GOres[[3]]
resKS.elim <- GOres[[4]]
allRes <- GOres[[5]]
showSigOfNodes(GOdata, score(resKS.elim), firstSigNodes = 5, useInfo = 'all')

```
# GSEA stuff

```{r GSEA}
### For GSEA Pre-ranked
htmp <- df_human[order(df_human$symbol, -df_human$dc_score),]
htmp2 <- htmp[ !duplicated(htmp$symbol),]
hprernk_dc <- data.frame(name = htmp2$symbol,
                         metric = htmp2$dc_score*-log10(htmp2$pval))

mtmp <- df_mouse[order(df_mouse$symbol, -df_mouse$dc_score),]
mtmp2 <- mtmp[ !duplicated(mtmp$symbol),]
mprernk_dc <- data.frame(name = mtmp2$symbol,
                         metric = mtmp2$dc_score*-log10(mtmp2$pval))

# write.table(hprernk_dc, file=file.path("../Data/hprernk_dc.rnk"),
#             row.names=FALSE, sep="\t", quote=FALSE)
# write.table(mprernk_dc, file=file.path("../Data/mprernk_dc.rnk"),
#             row.names=FALSE, sep="\t", quote=FALSE)

### Mouse and human non-SLE gene set overlap
hctl_file <- file.path("../GSEA/human_DCprerank/gsea_report_for_na_neg_1620863432696.tsv")
mctl_file <- file.path("../GSEA/mouse_DCprerank/gsea_report_for_na_neg_1620862959177.tsv")
hgsea_ctl <- read.delim(hctl_file) # 12 sets
mgsea_ctl <- read.delim(mctl_file) # 0 sets

hgsea_ctl05 <- hgsea_ctl[which(hgsea_ctl$FDR.q.val<0.25 & hgsea_ctl$NOM.p.val<0.05),
                         c("NAME","SIZE","NES","NOM.p.val","FDR.q.val")]
mgsea_ctl05 <- mgsea_ctl[which(mgsea_ctl$FDR.q.val<0.25 & mgsea_ctl$NOM.p.val<0.05),
                         c("NAME","SIZE","NES","NOM.p.val","FDR.q.val")]
ctl_overlap <- intersect(hgsea_ctl05$NAME, mgsea_ctl05$NAME)

### Mouse and human SLE gene set overlap
hsle_file <- file.path("../GSEA/human_DCprerank/gsea_report_for_na_pos_1620863432696.tsv")
msle_file <- file.path("../GSEA/mouse_DCprerank/gsea_report_for_na_pos_1620862959177.tsv")
hgsea_sle <- read.delim(hsle_file) # 761 sets
mgsea_sle <- read.delim(msle_file) # 674 sets

hgsea_sle05 <- hgsea_sle[which(hgsea_sle$FDR.q.val<0.25 & hgsea_sle$NOM.p.val<0.05),
                         c("NAME","SIZE","NES","NOM.p.val","FDR.q.val")]
mgsea_sle05 <- mgsea_sle[which(mgsea_sle$FDR.q.val<0.25 & mgsea_sle$NOM.p.val<0.05),
                         c("NAME","SIZE","NES","NOM.p.val","FDR.q.val")]
sle_overlap <- intersect(hgsea_sle05$NAME, mgsea_sle05$NAME)

### Dot plots
# Overlap; since 0 mouse ctl, skip that condition
ggplot(hgsea_sle05[hgsea_sle05$NAME %in% sle_overlap,], 
       aes(x=-log10(NOM.p.val),
           y=substring(NAME,10,50),
           size=SIZE,
           color=NES)) +
  geom_point() +
  scale_color_gradient(low = "blue", high = "red") +
  theme_bw() +
  ylab("Reactome Pathways") +
  xlab("-log10(pval)") +
  ggtitle("Overlapping DC Pathways - Human SLE")
ggplot(mgsea_sle05[mgsea_sle05$NAME %in% sle_overlap,], 
       aes(x=-log10(NOM.p.val),
           y=substring(NAME,10,50),
           size=SIZE,
           color=NES)) +
  geom_point() +
  scale_color_gradient(low = "blue", high = "red") +
  theme_bw() +
  ylab("Reactome Pathways") +
  xlab("-log10(pval)") +
  ggtitle("Overlapping DC Pathways - Mouse DKO")

```
# Visualization

```{r visualize functional enrichment}
### Mouse dotplot
df_dot <- mallRes[,c("Term","elimKS")]
df_dot <- df_dot[1:20,]
dotchart(as.numeric(df_dot$elimKS), labels=df_dot$Term, cex=.7,
         main="Mouse DCG enrichment for GO terms", xlab="KS elim pval")

### Human dotplot
df_dot <- hallRes[,c("Term","elimKS")]
df_dot <- df_dot[1:20,]
dotchart(as.numeric(df_dot$elimKS), labels=df_dot$Term, cex=.7,
         main="Human DCG enrichment for GO terms", xlab="KS elim pval")

### Overlap dotplot
df_dot <- allRes[,c("Term","elimKS")]
df_dot <- df_dot[1:20,]
dotchart(as.numeric(df_dot$elimKS), labels=df_dot$Term, cex=.7,
         main="Overlap DCG enrichment for GO terms", xlab="KS elim pval")

# Save dot plots to pdf
# had to manually save
```