---
title: "Differential Expression Analysis"
author: "Bri Ko"
date: "4/11/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# Load packages

```{r load, include=FALSE}
library(readr)
library(dplyr) # for %<% mutating
library(ggplot2) # for plotting
library(ggrepel) # for plot text spacing
library(Rgraphviz) # for visualizing GO graphs
library(gridExtra) # for printing table as pdf
library(limma)
library(edgeR)
library(topGO)
set.seed(1234)  # random seed for reproducibility
```

# Download and clean data

```{r GEO data from downloaded csv}
# same as PCA_heatmap.Rmd code w/o normalized counts (e.g. mX, hX)
### Mouse data
# Import the data and pull out gene names, counts (log2 cpm)
mdata <- read.csv(file = '../Data/ABC_WT_DKO.csv')
mgenes <- mdata[,"genes"]
mgenes <- mgenes[complete.cases(mgenes)]
cpm_cols <- c("WT_1_log2.cpm.","WT_2_log2.cpm.", 
              "DKO_1_log2.cpm.","DKO_2_log2.cpm.")
mcounts <- mdata[mdata$genes %in% mgenes, cpm_cols]
rownames(mcounts) <- mgenes
mcounts <- t(mcounts) # samples x genes
mcounts <- 2^mcounts # retrieve raw cpm values
mID <- sub("\\_.*", "", rownames(mcounts)) # list of sample ptypes

### Repeat for human data (in fpkm! raw counts = fpkm*transcipt_length...)
hdata <- read.csv(file = '../Data/Human_SLE_Ctrl.csv')
hgenes <- hdata[,c("ENTREZID","SYMBOL")]
hgenes <- hgenes[complete.cases(hgenes),]
hcounts <- hdata[hdata$ENTREZID %in% hgenes$ENTREZID,4:ncol(hdata)]
hcounts <- hcounts[-which(grepl("2-Mar",hgenes[,2])),]
hgenes <- hgenes[-which(grepl("2-Mar",hgenes[,2])),]
rownames(hcounts) <- hgenes[,1]
hcounts <- t(hcounts)
hID <- sub("\\..*", "", rownames(hcounts))
```
# Functions

```{r Useful functions}
empiricalBayes <- function(cnts, ID, contrast, p_cutoff, fc_cutoff, topN) {
  ### Input:
  #     cnts: count data (samples x genes)
  #     ID: list of sample ptypes
  #     contrast: string of contrast formula (e.g. "DKO-WT" or "SLE-HC")
  #     p_cutoff: p-value threshold for significance
  #     fc_cutoff: abs(log fold change) threshold for significance
  #     topN: number of top genes for annotation
  
  ### Output:
  #     res: DE analysis results 
  
  # Create identification matrix mapping samples with ptype
  mat <- model.matrix(~0+ID)
  colnames(mat) <- substr(colnames(mat),3,nchar(colnames(mat)))
  
  # Fit a linear model of gene counts by the design matrix, so each group is compared. 
  # Transpose cpm counts to align with design matrix.
  model <- lmFit(t(cnts), mat)
  
  # Create contrast matrix
  cont_mat <- makeContrasts(contrasts=contrast, levels=mat)
  
  # Compute estimated coefficients from linear model fit, contrasts
  model2 <- contrasts.fit(model, cont_mat)
  
  # Takes microarray linear fit model, returns t-statistics, F-statistics, and 
  # log-odds of differential expression using empirical Bayes' moderation
  model2 <- eBayes(model2)
  
  # Check how many genes are significantly over-expressed or under-expressed
  # between conditions, and which are equally expressed
  table(decideTests(model2)) #returns # of over (+1) or under (-1) expressed genes
  
  # Get table of top-ranked genes, make gene names explicit variable
  res <- topTable(model2, number=Inf, adjust.method="BH")
  res <- tibble::rownames_to_column(res,"ID")
  
  #Reformat full results to have additional variables of interest, filter for 
  #set parameters, plot volcano plot of DE and significance
  res <- res %>% 
    mutate(Significant = adj.P.Val < p_cutoff, abs(logFC) > fc_cutoff ) %>% 
    mutate(Rank = 1:n(), Label = ifelse(Rank < topN, ID,"")) %>%
    mutate(logpval = -log10(adj.P.Val))

  return(res)
}

GOenrich <- function(geneList, genes, topGO_args) {
  ### Input
  #     geneList: DE analysis result values for enrichment analysis (e.g. pval)
  #     genes: list of genes of type specified in ID
  #     topGO_args: arguments for new topGOdata object
  #       - ontology: what kind of ontology to use (e.g. BP)
  #       - geneSelectFun: function to select significant genes??? INCORPORATE PVAL<0.05 HERE?
  #       - mapping: string of what database to use for mapping
  #       - ID: type of gene ID (e.g. symbol, entrez)
  
  ### Output
  #     ???: GO enriched terms and stats
  
  names(geneList) <- genes
  
  # Create topGOdata object
  GOdata <- new("topGOdata",
                ontology = go_args[[1]],
                allGenes = geneList,
                geneSelectionFun = go_args[[2]],
                annot = annFUN.org, 
                mapping = go_args[[3]],
                ID = go_args[[4]])
  
  # Run several kinds of enrichment tests
  resFisher <- runTest(GOdata, algorithm = "classic", statistic = "fisher")
  resKS <- runTest(GOdata, algorithm = "classic", statistic = "ks")
  resKS.elim <- runTest(GOdata, algorithm = "elim", statistic = "ks")
  
  # Generate table of results of enrichment results
  allRes <- GenTable(GOdata, classicFisher = resFisher,
                     classicKS = resKS, elimKS = resKS.elim,
                     orderBy = "elimKS", ranksOf = "classicFisher", 
                     topNodes = 20)
  
  res <- list(GOdata, resFisher, resKS, resKS.elim, allRes)
  return(res)
}
```

# Differential Expression Analysis ("DE")

```{r Differential Expression}
### Mouse
mres <- empiricalBayes(mcounts, mID, contrast="DKO-WT", 
                       p_cutoff=0.05, fc_cutoff=1, topN=0.2*dim(mcounts)[2])

### Human
hres <- empiricalBayes(hcounts, hID, contrast="SLE-HC", 
                       p_cutoff=0.05, fc_cutoff=1, topN=0.2*dim(hcounts)[2])
```
# DEG overlap between human and mouse

```{r DE overlaps}
# Map human ENTREZ ids -> Symbol
h_entrez <- mapIds(org.Hs.eg.db, keys=hres$ID, column="SYMBOL", keytype="ENTREZID")
hres$Symbol <- h_entrez

# Identify up- and down-regulated genes in both species
m_updegs <- subset(mres, logFC > 0 & Significant == TRUE)
m_downdegs <- subset(mres, logFC < 0 & Significant == TRUE)
h_updegs <- subset(hres, logFC > 0 & Significant == TRUE)
h_downdegs <- subset(hres, logFC < 0 & Significant == TRUE)

# Identify overlapping genes in both species
up_overlap <- intersect(toupper(m_updegs$ID), h_updegs$Symbol)
down_overlap <- intersect(toupper(m_downdegs$ID), h_downdegs$Symbol)
overlap <- c(up_overlap, down_overlap)

# Convert GO terms in publicly available db to a list?
human_GO_terms <- org.Hs.egGO
mapped_genes <- mappedkeys(human_GO_terms) # entrez:GO mapping
human_GO_list <- as.list(human_GO_terms[mapped_genes])
# GET OVERLAPPING GENES' ASSOCIATED GO TERMS?

```
# GO Enrichment of DEGs using topGO

``` {GO enrichment DEGs}
# GO enrichment of DEGs
# documentation: https://www.bioconductor.org/packages/devel/bioc/vignettes/topGO/inst/doc/topGO.pdf
# walkthru: https://ucdavis-bioinformatics-training.github.io/2018-June-RNA-Seq-Workshop/friday/enrichment.html
### Mouse
# Set up input arguments
geneSelectFun <- function(x)x
go_args <- c("BP", geneSelectFun, "org.Mm.eg.db", "SYMBOL")

# Output GO enrichment results
mGOres <- GOenrich(mres$adj.P.Val, mres$ID, go_args) 
mGOdata <- mGOres[[1]]
mresFisher <- mGOres[[2]]
mresKS <- mGOres[[3]]
mresKS.elim <- mGOres[[4]]
mallRes <- mGOres[[5]]
showSigOfNodes(mGOdata, score(mresKS.elim), firstSigNodes = 5, useInfo = 'all')

### Human
# Set up input arguments
go_args <- c("BP", geneSelectFun, "org.Hs.eg.db", "ENTREZ")

# Output GO enrichment results as c(GOdata, resFisher, resKS, resKS.elim, allRes)
hGOres <- GOenrich(hres$adj.P.Val, hres$ID, go_args) 
hGOdata <- hGOres[[1]]
hresFisher <- hGOres[[2]]
hresKS <- hGOres[[3]]
hresKS.elim <- hGOres[[4]]
hallRes <- hGOres[[5]]
showSigOfNodes(hGOdata, score(hresKS.elim), firstSigNodes = 5, useInfo = 'all')
```
# GO Enrichment of overlapping genes using topGO

``` {GO enrichment overlap}
# Set up input arguments and subset human data for overlap
geneSelectFun <- function(x)x
go_args <- c("BP", geneSelectFun, "org.Hs.eg.db", "SYMBOL")
hres_overlap <- hres[which(hres$Symbol %in% overlap),]

# Output GO enrichment results
GOres <- GOenrich(hres_overlap$adj.P.Val, hres_overlap$Symbol, go_args) 
GOdata <- GOres[[1]]
resFisher <- GOres[[2]]
resKS <- GOres[[3]]
resKS.elim <- GOres[[4]]
allRes <- GOres[[5]]
showSigOfNodes(GOdata, score(resKS.elim), firstSigNodes = 5, useInfo = 'all')

```
# Visualization

```{r visualize functional enrichment}
### Mouse dotplot
df_dot <- mallRes[,c("Term","elimKS")]
df_dot <- df_dot[1:20,]
dotchart(as.numeric(df_dot$elimKS), labels=df_dot$Term, cex=.7,
         main="Mouse enrichment for GO terms", xlab="KS elim pval")

### Human dotplot
df_dot <- hallRes[,c("Term","elimKS")]
df_dot <- df_dot[1:20,]
dotchart(as.numeric(df_dot$elimKS), labels=df_dot$Term, cex=.7,
         main="Human enrichment for GO terms", xlab="KS elim pval")

### Overlap dotplot
df_dot <- allRes[,c("Term","elimKS")]
df_dot <- df_dot[1:20,]
dotchart(as.numeric(df_dot$elimKS), labels=df_dot$Term, cex=.7,
         main="Mouse-Human overlap enrichment for GO terms", xlab="KS elim pval")

# Save dot plots to pdf
# had to manually save
```